### PR TITLE
fix: delete hash entries instead of assigning null

### PR DESCRIPTION
**Pull request by dev7355608**
##### Description of change

If we don't delete, we'll have `null` entries for destroyed objects forever since the uids are not recycled, I think.

##### Pre-Merge Checklist

- [ ] Tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] Lint process passed (`npm run lint`)
- [ ] Tests passed (`npm run test`)

### CODE REVIEW

File: src/rendering/batcher/shared/BatchTextureArray.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ export class BatchTextureArray
             const t = this.textures[i];
 
             this.textures[i] = null;
-            this.ids[t.uid] = null;
```
Comment (GoodBoyDigital) [inline src/rendering/batcher/shared/BatchTextureArray.ts:32]: you are not wrong, but delete is a suuppperr slow operation

Comment (dev7355608) [reply src/rendering/batcher/shared/BatchTextureArray.ts:32]: Yeah, but I imagine if these object have tens of thousands of unnecessary entries that can't be good either. In theory you could probably run out of memory.

Comment (GoodBoyDigital) [reply src/rendering/batcher/shared/BatchTextureArray.ts:32]: maybe there is like a middle ground - could just reset them every X frames ?

Comment (GoodBoyDigital) [reply src/rendering/batcher/shared/BatchTextureArray.ts:32]: My main concern is that temoving keys changes the object's shape and can lead to de-optimization.

**Review by GoodBoyDigital — CHANGES_REQUESTED** [review]
Had a think about this - as this is a common thing (especially across systems).

lets use the `SchedulerSystem` to periodically run the following every 60 secs?:

```
const objectToClean

cont newObject = Object.create(null);

for(var i in objectToClean)
{
    const item = objectToClean[i]
    if(item)newObject[i] = item;
}

return newObject;
```

what do you think @dev7355608 ?

for the `BatchTextureArray` thats SUPER hot code. It should not get too big as its only using textures which are generaly fairly stable. Will have a think though - we could do something similar, like periodically clean the batchers...

**Comment by dev7355608** [comment]
I was wondering about Object vs Map. I assume it was benchmarked for these integer key records? I think recall Maps performing better for this kind of use case, especially with a lot of entries. I can't imagine this being great for performance:
```js
for (const i in object)
```
because in theory either the integer key was already converted to a string at the time the entry was put in the object or it is converted when we iterate it. I'm not sure if the optimizer is so smart that no string is created here.
With Maps we don't have this problem that our keys are strings.

**Comment by dev7355608** [comment]
Regarding the `BatchTextureArray`: Considering the number or textures is small, perhaps iterating over the array instead is just as fast that the object access and so we don't need `ids` at all?

```ts
    public getTextureId(texture: TextureSource): number
        const uid = texture.uid;
        const textures = this.textures;
        const count = this.count;

        for (let i = 0; i < count; i++)
        {
            if (textures[i].uid === uid) return i;
        }

        return -1;
    }
```

**Comment by GoodBoyDigital** [comment]
> I was wondering about Object vs Map. I assume it was benchmarked for these integer key records? I think recall Maps performing better for this kind of use case, especially with a lot of entries. I can't imagine this being great for performance:
> 
> ```js
> for (const i in object)
> ```
> 
> because in theory either the integer key was already converted to a string at the time the entry was put in the object or it is converted when we iterate it. I'm not sure if the optimizer is so smart that no string is created here. With Maps we don't have this problem that our keys are strings.

Maps don't tend to cause de-optimisations which is great, but annoyingly reading from them is waaay slower than reading from a standard object..

Every PR creates a bunnymark sandbox now, so it should at least be easy to test this stuff!

**Comment by GoodBoyDigital** [comment]
> Regarding the `BatchTextureArray`: Considering the number or textures is small, perhaps iterating over the array instead is just as fast that the object access and so we don't need `ids` at all?
> 
> ```ts
>     public getTextureId(texture: TextureSource): number
>         const uid = texture.uid;
>         const textures = this.textures;
>         const count = this.count;
> 
>         for (let i = 0; i < count; i++)
>         {
>             if (textures[i].uid === uid) return i;
>         }
> 
>         return -1;
>     }
> ```

could work, we can bench mark this one for sure!

**Comment by GoodBoyDigital** [comment]
hey, had another thought for pipes.. we could attach the objects directly to the renderable itself (eg Sprite / Graphics)
this could be done by adding some kind of `_gpuRenderCache` typed as `unknown` and then the pipe can declare the type it wants. 

That way we don't need the majority of the hashes and we still keep the Renderabels agnostic to Canvas/WebGPU/WebGL?

What doya think?

**Comment by dev7355608** [comment]
The `_gpuRenderCache` property sounds good. But then you can't render the the same container in different renderers. But I'm not sure that is something that is actually supported. Just I thought I had.

**Comment by GoodBoyDigital** [comment]
heya @dev7355608 followed up with this PR #10793

I tried adding the items directly to the sprite - and for some reason its slower?? Perhaps many objects = slower than a single large one?

With that in mind i wend for having the hashes periodically cleaned. I will open an issue for the BatchTextureArray. This one is in the heart of the rendering loop, so keen to find a solution that does not compromise on perf!

**Comment by GoodBoyDigital** [comment]
closing in favour of #10793 . Thanks for identifying and providing the initial fix for this one @dev7355608 !!

**Pull request closed without merging**